File: certification_server/src/iface/authcontrol/authcontrol.go
package authcontrol

import (
	authmodel "certification_server/src/models/auth"
	modelsystem "certification_server/src/models/system"
	"context"
)

// InboundRateLimitInput 是 authcontrol 使用的与协议无关的入站输入。
type InboundRateLimitInput struct {
	Scope     authmodel.RateLimitScope
	Transport string

	Module string
	Action string
	Route  string
	Method string

	SourceIP  string
	GatewayID string
	ClientID  string

	SourceService string
	TargetService string

	Headers map[string]string
	Tags    map[string]string

	Identity *authmodel.IdentityContext
}

// Build 将入站输入转换为统一的限流描述符。
func Build(input *InboundRateLimitInput) (*authmodel.RateLimitDescriptor, error) {
	if input == nil {
		return nil, &modelsystem.ErrInboundRateLimitInputNil
	}

	descriptor := &authmodel.RateLimitDescriptor{
		Scope:         input.Scope,
		Transport:     input.Transport,
		Module:        input.Module,
		Action:        input.Action,
		Route:         input.Route,
		Method:        input.Method,
		SourceIP:      input.SourceIP,
		GatewayID:     input.GatewayID,
		ClientID:      input.ClientID,
		SourceService: input.SourceService,
		TargetService: input.TargetService,
		Tags:          input.Tags,
	}
	if descriptor.Tags == nil {
		descriptor.Tags = map[string]string{}
	}

	if input.Identity != nil {
		descriptor.Authenticated = true
		descriptor.EntityType = input.Identity.EntityType
		descriptor.EntityID = input.Identity.EntityID
		descriptor.PrincipalID = input.Identity.PrincipalID
		descriptor.SessionID = input.Identity.SessionID.String()
		descriptor.TokenID = input.Identity.TokenID.String()
		descriptor.TokenType = input.Identity.TokenType
		descriptor.Scopes = input.Identity.Scopes
	}

	return descriptor, nil
}

// InboundControlRequest 是认证中心本地入站控制请求。
type InboundControlRequest struct {
	RateLimitInput *InboundRateLimitInput
}

// InboundControlResult 是认证中心本地入站控制结果。
type InboundControlResult struct {
	RateLimitDecision *authmodel.RateLimitDecision
}

// IInboundAuthControl 定义 certification_server 的本地入站认证控制。
//
// 说明：
//   - 该门面只暴露入站控制结果，不再拆分独立的 rate-limit 能力接口层。
//   - 描述符构建与限流决策属于内部实现细节，不应向上层继续分叉。
type IInboundAuthControl interface {
	EnforceInbound(ctx context.Context, req *InboundControlRequest) (*InboundControlResult, error)
}

File: data_worker/src/iface/authcontrol/auth_control.py
from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from typing import cast

from src.models.auth.auth import IdentityContext
from src.models.auth.ratelimit import RateLimitScope
from src.models.auth.auth import TokenType
from src.models.auth.ratelimit import RateLimitDecision, RateLimitDescriptor
from src.models.common.entry_type import EntityType


@dataclass(slots=True, kw_only=True)
class InboundRateLimitInput:
    scope: RateLimitScope
    transport: str

    module: str
    action: str
    route: str
    method: str

    source_ip: str = ""
    gateway_id: str = ""
    client_id: str = ""

    source_service: str = ""
    target_service: str = ""

    headers: dict[str, str] = field(default_factory=dict)
    tags: dict[str, str] = field(default_factory=dict)

    identity: IdentityContext | None = None


def Build(input_data: InboundRateLimitInput | None) -> RateLimitDescriptor:
    if input_data is None:
        raise ValueError("inbound rate limit input is nil")

    descriptor = RateLimitDescriptor(
        scope=input_data.scope,
        transport=input_data.transport,
        module=input_data.module,
        action=input_data.action,
        route=input_data.route,
        method=input_data.method,
        authenticated=False,
        source_ip=input_data.source_ip,
        gateway_id=input_data.gateway_id,
        client_id=input_data.client_id,
        source_service=input_data.source_service,
        target_service=input_data.target_service,
        entity_type="unknown",
        entity_id="",
        principal_id="",
        session_id="",
        token_id="",
        token_type=cast(TokenType, ""),
        scopes=[],
        tags=dict(input_data.tags or {}),
    )

    identity = input_data.identity
    if identity is not None:
        descriptor.authenticated = True
        descriptor.entity_type = identity.entity_type
        descriptor.entity_id = identity.entity_id
        descriptor.principal_id = identity.principal_id
        descriptor.session_id = str(identity.session_id)
        descriptor.token_id = str(identity.token_id)
        descriptor.token_type = identity.token_type
        descriptor.scopes = list(identity.scopes)

    return descriptor


@dataclass(slots=True, kw_only=True)
class InboundControlRequest:
    rate_limit_input: InboundRateLimitInput | None = None


@dataclass(slots=True, kw_only=True)
class InboundControlResult:
    rate_limit_decision: RateLimitDecision | None = None


class IInboundAuthControl(ABC):
    """data_worker 的本地入站认证控制。"""

    @abstractmethod
    async def enforce_inbound(self, req: InboundControlRequest) -> InboundControlResult:
        raise NotImplementedError
